ui/ts/test/util.ts

```diff
@@ -1,9 +1,193 @@
+
+      test("computed epoch is always parent epoch.", () => {
+        for (let i = 1; i < 3; i++) {
+          let newVal = "new_" + i.toString();
+          parent(newVal);
+          assert.equal(parent.Epoch(), i);
+          assert.equal(computed.Epoch(), parent.Epoch());
+        }
+      });
+
+      test("computed property can depend on another computed property", () => {
+        let computed2 = Utils.Computed(computed, (s: string) => s + "+recomputed");
+        assert.equal(computed2(), "initial+computed1+recomputed");
+        parent("newvalue");
+        assert.equal(computed2(), "newvalue+computed2+recomputed");
+      });
+    });
+
+    suite("Multi-Parent", () => {
+      let parents: Utils.Property<string>[];
+      let recomputeCount: number;
+      let computed: Utils.ReadOnlyProperty<string>;
+      let computedString = (s1: string, s2: string, s3: string, s4: string, n: number) => {
+        return [s1, s2, s3].join(", ") + " and " + s4 + "(" + n + ")";
+      };
+
+      setup(() => {
+        parents = [
+          Utils.Prop("Leonardo"),
+          Utils.Prop("Donatello"),
+          Utils.Prop("Raphael"),
+          Utils.Prop("Michaelangelo")
+        ];
+        recomputeCount = 0;
+        computed = Utils.Computed(parents[0], parents[1], parents[2], parents[3],
+                                  (s1: string, s2: string, s3: string, s4: string) => {
+          recomputeCount++;
+          return computedString(s1, s2, s3, s4, recomputeCount);
+        });
+      });
+
+      test("computed value is lazily recomputed.", () => {
+        for (var i = 1; i < 3; i++) {
+          parents.forEach((p: Utils.Property<string>) => {
+            p(p() + "_" + i.toString());
+            assert.equal(p.Epoch(), i);
+          });
+          assert.equal(computed(), computedString(parents[0](), parents[1](), parents[2](), parents[3](), i));
+          assert.equal(computed(), computedString(parents[0](), parents[1](), parents[2](), parents[3](), i));
+        }
+        assert.equal(computed(), "Leonardo_1_2, Donatello_1_2, Raphael_1_2 and Michaelangelo_1_2(2)");
+      });
+
+      test("computed value is recomputed if only one parent changes.", () => {
+        for (var i = 1; i < 3; i++) {
+          parents[0](parents[0]() + "_" + i.toString());
+          assert.equal(computed(), computedString(parents[0](), parents[1](), parents[2](), parents[3](), i));
+          assert.equal(computed(), computedString(parents[0](), parents[1](), parents[2](), parents[3](), i));
+        }
+        assert.equal(computed(), "Leonardo_1_2, Donatello, Raphael and Michaelangelo(2)");
+      });
+
+      test("computed epoch is always sum of parent epochs.", () => {
+        for (var i = 1; i < 3; i++) {
+          parents.forEach((p: Utils.Property<string>) => {
+            p(p() + "_" + i.toString());
+            assert.equal(p.Epoch(), i);
+          });
+          assert.equal(computed.Epoch(), parents[0].Epoch() * parents.length);
```

Comment:
This test might be more effective if for the first loop you set parent 1, for the 2nd you set 1 and 2, etc.